encryption.md

```diff
@@ -0,0 +1,146 @@
+		if not success then
+			wezterm.log_error(stderr)
+		else
+			return stdout
+		end
+	end,
+})
+```
+
+# Providers
+If you think something is missing, then please provide a PR or an issue.
+
+## Rage
+[Rage](https://github.com/str4d/rage) is a drop in replacement for age and can be used by installing it to the path, and then using the following code:
+
+```lua
+local wezterm = require("wezterm")
+local resurrect = wezterm.plugin.require("https://github.com/MLFlexer/resurrect.wezterm")
+
+local function execute_shell_cmd(cmd)
+	local is_windows = wezterm.target_triple == "x86_64-pc-windows-msvc"
+	local process_args = is_windows and { "cmd.exe", "/C", cmd } or { os.getenv("SHELL"), "-c", cmd }
+	local success, stdout, stderr = wezterm.run_child_process(process_args)
+	return success, stdout, stderr
+end
+
+local public_key = "public_key"
+local private_key = "/path/to/private/key.txt"
+
+resurrect.set_encryption({
+  enable = true,
+  private_key = private_key,
+  public_key = public_key,
+	encrypt = function(file_path, lines)
+		local success, _, stderr = execute_shell_cmd(
+			string.format(
+				"echo %s | rage -r %s -o %s",
+				wezterm.shell_quote_arg(lines),
+				public_key,
+				file_path:gsub(" ", "\\ ")
+			)
+		)
+		if not success then
+			wezterm.log_error(stderr)
+		end
+	end,
+	decrypt = function(file_path)
+		local success, stdout, stderr =
+			execute_shell_cmd(string.format('rage -d -i "%s" "%s"', private_key, file_path))
+		if not success then
+			wezterm.log_error(stderr)
+		else
+			return stdout
+		end
+	end,
+})
+```
+## GPG
+[GnuPG](https://gnupg.org/) can be used by installing it to the path.
+Then generating a key pair: `gpg --full-generate-key`
+Get the public key with `gpg --armor --export your_email@example.com`
+
+
+```lua
+local wezterm = require("wezterm")
+local resurrect = wezterm.plugin.require("https://github.com/MLFlexer/resurrect.wezterm")
+
+local function execute_shell_cmd(cmd)
+	local is_windows = wezterm.target_triple == "x86_64-pc-windows-msvc"
+	local process_args = is_windows and { "cmd.exe", "/C", cmd } or { os.getenv("SHELL"), "-c", cmd }
+	local success, stdout, stderr = wezterm.run_child_process(process_args)
+	return success, stdout, stderr
+end
+
+local public_key = "your_email@example.com"
+
+resurrect.set_encryption({
+  enable = true,
+  public_key = public_key,
+  encrypt = function(file_path, lines)
+    local success, _, stderr = execute_shell_cmd(
+      string.format(
+        'echo %s | gpg --encrypt --recipient %s --output %s',
+        wezterm.shell_quote_arg(lines),
+        public_key,
+        file_path:gsub(" ", "\\ ")
+      )
+    )
+    if not success then
+      wezterm.log_error(stderr)
+    end
+  end,
+  decrypt = function(file_path)
+    local success, stdout, stderr =
+			execute_shell_cmd(string.format("gpg --decrypt \"%s\" | tr -d '\\000-\\037'", file_path)) -- TODO: fix (\u0000-\u001F) from being included
```

Comment:
There is a problem with chars in the range (\u0000-\u001F) being included in the json, this might not be caused by the decryption, but rather the encryption. Adding `| tr -d '\\000-\\037'` makes it possible to restore some of the state, but it is not possible to restore completely as some of the state has been corrupted by removing the `(\u0000-\u001F)`.

This will need more investigation

Reply:
---@param file_path string
---@param state table
local function write_state(file_path, state)
	local json_state = wezterm.json_encode(state)
	json_state = json_state:gsub("[%c]", "") -- remove control characters
	if pub.encryption.enable then
		pub.encryption.encrypt(file_path, json_state)
	else
		local file = assert(io.open(file_path, "w"))
		file:write(json_state)
		file:close()
	end
end

---@param file_path string
---@return table
local function load_json(file_path)
	if pub.encryption.enable then
		local json = pub.encryption.decrypt(file_path)
		if not json then
			return {}
		end
		json = json:gsub("[%c]", "") -- remove control characters
		return wezterm.json_parse(json)
	else
		local lines = {}
		for line in io.lines(file_path) do
			table.insert(lines, line)
		end
		local json = table.concat(lines)
		return wezterm.json_parse(json)
	end
end

I ended up with this error and was able to avoid it with the above commented lines to sanitize control chars from the json.

Reply:
I came across another error with improperly sanitized UTF8 chars.
We may need a `sanitize_json()` function:
```lua
--- Sanitize the input by removing control characters and invalid UTF-8 sequences
--- @param data string
--- @return string
local function sanitize_json(data)
    -- Remove control characters (except for \n, \r, \t)
    data = data:gsub("[%c]", function(c)
        if c == "\n" or c == "\r" or c == "\t" then
            return c
        end
        return ""
    end)
    -- Remove invalid UTF-8 sequences
    return data:gsub("[^\x00-\x7F]", "")
end
```

Reply:
Awesome that you investigated this! and made a PR, really appreciate your contributions :smile: 

Did you figure out the cause of these issues, I did not notice these errors before trying GPG, so i suspect it to be caused by the GPG encryption?